File: leetcode/2391/solution.go
/**
 * https://leetcode.com/problems/minimum-amount-of-time-to-collect-garbage/
 * 2391. Minimum Amount of Time to Collect Garbage
 */
package main

func garbageCollection(garbage []string, travel []int) int {
	var total, M, P, G int

	for i := 1; i < len(travel); i++ {
		travel[i] += travel[i-1]
	}

	for i, s := range garbage {
		for _, x := range s {
			switch x {
			case 'M':
				M = i
			case 'P':
				P = i
			default:
				G = i
			}
			total++
		}
	}

	if M > 0 {
		total += travel[M-1]
	}
	if P > 0 {
		total += travel[P-1]
	}
	if G > 0 {
		total += travel[G-1]
	}

	return total
}

File: leetcode/2391/solution.rs
/**
 * https://leetcode.com/problems/minimum-amount-of-time-to-collect-garbage/
 * 2391. Minimum Amount of Time to Collect Garbage
 */

impl Solution {
    pub fn garbage_collection(garbage: Vec<String>, mut travel: Vec<i32>) -> i32 {
        let mut total = 0;
        let (mut mi, mut pi, mut gi) = (0, 0, 0); // last index of each garbage

        for i in 1..travel.len() {
            travel[i] += travel[i - 1];
        }

        for (i, g) in garbage.iter().enumerate() {
            for c in g.chars() {
                match c {
                    'M' => mi = i,
                    'P' => pi = i,
                    _ => gi = i,
                }
                total += 1;
            }
        }

        if mi != 0 {
            total += travel[mi - 1];
        }
        if pi != 0 {
            total += travel[pi - 1];
        }
        if gi != 0 {
            total += travel[gi - 1];
        }

        total
    }
}
